docs/determinism.md

```diff
@@ -0,0 +1,84 @@
+# Determinism 🤝 Rollups
+
+[Determinism](https://en.wikipedia.org/wiki/Deterministic_system#In_computer_science) is crucial for rollups because of [fault proof](https://community.optimism.io/docs/protocol/2-rollup-protocol/#fault-proofs). Validators will execute program([Turing machine](https://en.wikipedia.org/wiki/Turing_machine) tape) to validate the proposed state of rollup, syncing shared [repliced state machine](https://en.wikipedia.org/wiki/State_machine_replication). The state machine consists of a transition function. The transition function must be deterministic to decide which party is correct when dispute occur. For example, let's say that Alice is the state proposer and Bob is a validator. Bob will advance its own machine state by validating the program, and checks the sanity of proposed state at L1 and disagrees with it, and tells L1 that Alice has lied, triggering fault proof. If the program is nondeterministic, L1 will never find out who's claim(state) is right, leading L1 security inheritance loss.
+
+<p align="center">
+  <img src="assets/dispute.png" />
+</p>
+
+Note that determinism is only required when validating. When sequencer executes transactions(not validating), determinism of the entire program is not required. EVM state machine will provide necessary determinism for execution. For example, let's say that the sequencer uses [geth](https://github.com/ethereum/go-ethereum)'s EVM implementation for transaction execution. Even if geth program is nondeterministic because of parallelism(multiprocessing + goroutines), EVM state machine will still be deterministic. When validating, since required logic for validation is a part of execution program, the required logic can be taken care for guaranteeing determinism, such as compiling down to [Wasm](https://en.wikipedia.org/wiki/WebAssembly) or [MIPS](https://en.wikipedia.org/wiki/MIPS_architecture). 
+
+Every validator must share identical virtual machine structure for supporting determinism. If not, nondeterminism will sneak in. For example, Alice is validating MIPS program using virtual machine where pipelining is disabled. Bob is validating using actual MIPS chip which pipelining is enabled. Intermediate state may be different due to discrepancy of validation machine.
+
+There are multiple causes which introduces nondeterminism, such as parallelism, concurrency, virtual machine difference and even instruction level nondeterminism. We need to dodge this issues while validating(proving), making the proved program be deterministic. Let's observe how the two big rollup giants: [Arbitrum](https://arbitrum.io/) and [Optimism](https://www.optimism.io/), managed to overcome nondeterminism and handled side effects. After that, we will discuss food for thought topics related with determinism.
```

Comment:
```suggestion
To guarantee that a rollup inherits L1's security, we should ensure that Alice and Bob's clients generate the same value as the on-chain validation program. On-chain validation program could easily guarantee determinism because it's executed on EVM, where determinism is built-in. However, Alice and Bob's off-chain clients may not be that deterministic depending on how their clients solved the determinism problem.

Guaranteeing deterministic execution for off-chain clients is quite challenging since it's hard to build an execution environment that works simultaneously at the machine-language level for every computer. We should resolve some difficult issues such as parallelism, concurrency, virtual machine difference, system call, and non-deterministic machine instructions. 

Let's take Alice and Bob's example, where they are both validators and Alice uses an 8-core CPU, and Bob has a 4-core CPU. Even though they are running the same program, machine-level instruction cannot be identical since the number of cores is different. Not to mention, it's not guaranteeing deterministic execution. Many similar issues destroy deterministic execution in the computing world.

Arbitrum and Optimism solved the determinism problem and achieved their mission despite its difficulty. In this article, we'll figure out how each two rollup giants solved the problem, handled side effects, and some food for thoughts related to determinism.
```

Not sure if it's good though. Let's chat tmr.